# Indexing and selecting

In [1]:
# create dataframe
import pandas as pd
import numpy as np
my_list = np.array([[1,2,4],[4,5,6]])
df = pd.DataFrame(data=my_list,columns=list("xyz"),index=list("ab"))
df


x  y  z
a  1  2  4
b  4  5  6

In [2]:
# create dataframe
import pandas as pd
import numpy as np
my_list = np.array([[1,2,4],[4,5,6]])
df = pd.DataFrame(data=my_list,columns=['col1','col2','col3'])
df


col1  col2  col3
0     1     2     4
1     4     5     6

In [4]:
# kreye yon application ki simile note etidyan yo pou boutcamp lan
columns = ['Python','EDA','DM','BA','AS','ANS',"CAP"]
index =['FLO','WIZ','BON','GOT','KES','WED','NANA','FRE','SLY','DEZ','CAM','PYE','SEJ','DAV','PAS','CHRIS','ADI','CHAS','MES','DJE']
note = np.random.randint(55,100,(len(index),len(columns)))


In [5]:
df = pd.DataFrame(data=note,index=index,columns=columns)
df["Moy"]=np.round(df.mean(axis=1),2)
df.sort_values(by="Moy",ascending=0)

Python  EDA  DM  BA  AS  ANS  CAP    Moy
DJE        73   88  96  84  77   80   83  83.00
NANA       81   62  78  93  83   85   93  82.14
ADI        73   66  92  91  79   86   87  82.00
CHRIS      91   79  82  74  64   85   93  81.14
GOT        87   82  75  75  68   85   94  80.86
PAS        74   85  93  68  59   97   81  79.57
SLY        99   79  62  56  91   97   71  79.29
BON        75   84  84  85  61   72   88  78.43
PYE        80   68  73  96  82   67   78  77.71
MES        72   98  68  55  68   92   89  77.43
CAM        68   94  70  90  69   70   81  77.43
DEZ        86   56  87  78  99   55   81  77.43
WED        76   80  83  86  68   76   73  77.43
KES        97   89  58  92  64   80   60  77.14
CHAS       62   78  61  68  82   98   89  76.86
SEJ        75   86  86  58  66   83   73  75.29
DAV        65   63  73  83  83   70   89  75.14
FRE        62   95  95  69  63   58   68  72.86
WIZ        79   61  72  71  69   71   73  70.86
FLO        83   59  92  55  71   68   62  70.00

In [37]:
values=dict(a=[1,3,4,3],b=[1,2,4,4],c=[1,3,4,5])

df = pd.DataFrame(values)
df

a  b  c
0  1  1  1
1  3  2  3
2  4  4  4
3  3  4  5

In [48]:
values=[{"one":1,"two":3,"three":5},{"one":4,"two":3.5,"three":5.5},{"one":4,"two":3,"four":9}]
df =pd.DataFrame(values)
df.tail(5)

one  two  three  four
0    1  3.0    5.0   NaN
1    4  3.5    5.5   NaN
2    4  3.0    NaN   9.0

In [42]:
values=[(("a",1),("b",2)),(("a",4),("b",5))]
df =pd.DataFrame(values)
df[0][0][1]

1

In [1]:
import pandas as pd
food=pd.read_excel("sampledatafoodsales.xlsx", sheet_name=1)

Obje natif yo nan Python ban nou bon mwayen pou n endekse done yo. Pandas erite de tout sa yo. Ann konsidere DataFrame sa:

In [2]:
food

OrderDate Region         City  Category         Product  Quantity  \
0   2020-01-01   East       Boston      Bars          Carrot        33   
1   2020-01-04   East       Boston  Crackers     Whole Wheat        87   
2   2020-01-07   West  Los Angeles   Cookies  Chocolate Chip        58   
3   2020-01-10   East     New York   Cookies  Chocolate Chip        82   
4   2020-01-13   East       Boston   Cookies       Arrowroot        38   
..         ...    ...          ...       ...             ...       ...   
239 2021-12-18   East       Boston   Cookies       Arrowroot        34   
240 2021-12-21   East       Boston   Cookies  Chocolate Chip       245   
241 2021-12-24   East       Boston  Crackers     Whole Wheat        30   
242 2021-12-27   West  Los Angeles      Bars            Bran        30   
243 2021-12-30   West  Los Angeles   Cookies  Oatmeal Raisin        44   

     UnitPrice  TotalPrice  
0         1.77       58.41  
1         3.49      303.63  
2         1.87      108.46  
3         1.87      153.34  
4         2.18       82.84  
..         ...         ...  
239       2.18       74.12  
240       1.87      458.15  
241       3.49      104.70  
242       1.87       56.10  
243       2.84      124.96  

[244 rows x 8 columns]

Nan Python, nou kapab aksede a propriyete yon obje menm jan ak yon atribi. Pa egzanp, yon obje `moun` ka gen propriyete `wote`, pou nou aksede a wote moun lan nou kapab fe `moun.wote`. Se menm bagay la nan Pandas, pou nou aksede a yon kolonn, nou ka fe `dataframe.column`. Pa egzanp, pou nou aksede a kolonn `Region` an, nou kapab fe :

In [3]:
food.Region

0      East
1      East
2      West
3      East
4      East
       ... 
239    East
240    East
241    East
242    West
243    West
Name: Region, Length: 244, dtype: object

Men jan tou, si w gen yon diksyone nan Python, ou kapab aksede a yon vale ladan l, le w itilize `[]`. Nan ka sa pou nou aksede a kolonn `Region` an, nou kapab fe :

In [4]:
food['Region']

0      East
1      East
2      West
3      East
4      East
       ... 
239    East
240    East
241    East
242    West
243    West
Name: Region, Length: 244, dtype: object

Pandas Series yo fonksyone tankou yon diksyone, nou kapab fasilman jwenn premye Region an le nou fe:

In [5]:
food['Region'][0]

'East'

## Index

`Index` yo nan yon DataFrame ban nou yon mwayen pou nou idantifye liy yo.   
Ann regade DataFrame food la

In [6]:
food

OrderDate Region         City  Category         Product  Quantity  \
0   2020-01-01   East       Boston      Bars          Carrot        33   
1   2020-01-04   East       Boston  Crackers     Whole Wheat        87   
2   2020-01-07   West  Los Angeles   Cookies  Chocolate Chip        58   
3   2020-01-10   East     New York   Cookies  Chocolate Chip        82   
4   2020-01-13   East       Boston   Cookies       Arrowroot        38   
..         ...    ...          ...       ...             ...       ...   
239 2021-12-18   East       Boston   Cookies       Arrowroot        34   
240 2021-12-21   East       Boston   Cookies  Chocolate Chip       245   
241 2021-12-24   East       Boston  Crackers     Whole Wheat        30   
242 2021-12-27   West  Los Angeles      Bars            Bran        30   
243 2021-12-30   West  Los Angeles   Cookies  Oatmeal Raisin        44   

     UnitPrice  TotalPrice  
0         1.77       58.41  
1         3.49      303.63  
2         1.87      108.46  
3         1.87      153.34  
4         2.18       82.84  
..         ...         ...  
239       2.18       74.12  
240       1.87      458.15  
241       3.49      104.70  
242       1.87       56.10  
243       2.84      124.96  

[244 rows x 8 columns]

Nou kapab we ke chak liy gen yon nonb ki bo kote l, soti nan zero rive nan 243. Pa defo `index` la se lis nonb ki komanse nan zero a. Pa defo le w kreye yon DataFrame, si w pa tou spesifye `index` la. Pandas ap kreye youn otomatikman pou ou.  
Nan Pandas ou kapab idantifye liy yo avek led yon `label` oubyen pozisyon egzakt yo. Egzanp nou genyen pi wo a idantifye liy yo avek pozisyon egzakt yo,nou kapab itilize `label` tou, ann gade koman nou ka fe sa.


Pou nou asiye yon `index` ,nou itilize `set_index()`. Nou kapab pase l yon lis ki gen menm kantite eleman ak tablo a oubyen yon kolonn nan tablo a, ann eseye fe sa ak `food`.

In [7]:
food_label=food.set_index("OrderDate")
display(food_label)

Region         City  Category         Product  Quantity  UnitPrice  \
OrderDate                                                                       
2020-01-01   East       Boston      Bars          Carrot        33       1.77   
2020-01-04   East       Boston  Crackers     Whole Wheat        87       3.49   
2020-01-07   West  Los Angeles   Cookies  Chocolate Chip        58       1.87   
2020-01-10   East     New York   Cookies  Chocolate Chip        82       1.87   
2020-01-13   East       Boston   Cookies       Arrowroot        38       2.18   
...           ...          ...       ...             ...       ...        ...   
2021-12-18   East       Boston   Cookies       Arrowroot        34       2.18   
2021-12-21   East       Boston   Cookies  Chocolate Chip       245       1.87   
2021-12-24   East       Boston  Crackers     Whole Wheat        30       3.49   
2021-12-27   West  Los Angeles      Bars            Bran        30       1.87   
2021-12-30   West  Los Angeles   Cookies  Oatmeal Raisin        44       2.84   

            TotalPrice  
OrderDate               
2020-01-01       58.41  
2020-01-04      303.63  
2020-01-07      108.46  
2020-01-10      153.34  
2020-01-13       82.84  
...                ...  
2021-12-18       74.12  
2021-12-21      458.15  
2021-12-24      104.70  
2021-12-27       56.10  
2021-12-30      124.96  

[244 rows x 7 columns]

Nan egzanp sa nou itilize  kolonn `OrderDate` la kom `index`

Pou seleksyon done nan yon DataFrame Pandas genyen pwop operate li ki se `iloc[liy, kolonn]` avek `loc[liy, kolonn]`. Pou nou kapab fe operasyon ki pi avanse se yo nou dwe itilize.

## Seleksyon baze sou pozisyon nimerik

`iloc` pemet nou seleksyon done nan yon DataFrame avek led pozisyon nimerik li. Pa egzanp,pou nou seleksyone premye liy nan DataFrame `food` la, nou kapab fe:

In [8]:
food.iloc[0,:]

OrderDate     2020-01-01 00:00:00
Region                       East
City                       Boston
Category                     Bars
Product                    Carrot
Quantity                       33
UnitPrice                    1.77
TotalPrice                  58.41
Name: 0, dtype: object

Pou nou pran yon kolonn avek iloc, nou kapab fe:

In [9]:
food.iloc[:, 0] # Zero reprezante premye kolonn lan

0     2020-01-01
1     2020-01-04
2     2020-01-07
3     2020-01-10
4     2020-01-13
         ...    
239   2021-12-18
240   2021-12-21
241   2021-12-24
242   2021-12-27
243   2021-12-30
Name: OrderDate, Length: 244, dtype: datetime64[ns]

Le l poukont li, operate `:` vle di `pran tout vale`, men le w konbine l avek lot vale , li kapab pemet ou seleksyone yon ansanm vale. Pa egzanp, pou nou pran 3 premye liy nan premye kolonn lan nou ka fe:

In [10]:
food.iloc[:3, 0]

0   2020-01-01
1   2020-01-04
2   2020-01-07
Name: OrderDate, dtype: datetime64[ns]

## Seleksyon baze sou label

Dezyem fason ou ka seleksyone done nan yon DataFrame se le w itilize `label` , pou sa nou seve avec `loc`. Nan ekzanp food la nou kapab itilize:

In [11]:
food_label.loc['2020-01-01':'2020-01-10', :]

Region         City  Category         Product  Quantity  UnitPrice  \
OrderDate                                                                       
2020-01-01   East       Boston      Bars          Carrot        33       1.77   
2020-01-04   East       Boston  Crackers     Whole Wheat        87       3.49   
2020-01-07   West  Los Angeles   Cookies  Chocolate Chip        58       1.87   
2020-01-10   East     New York   Cookies  Chocolate Chip        82       1.87   

            TotalPrice  
OrderDate               
2020-01-01       58.41  
2020-01-04      303.63  
2020-01-07      108.46  
2020-01-10      153.34

Remake ke le nou itilize loc la, li pran denye vale nou spesifye a, se youn nan diferans ki genyen ant `iloc` avek `loc`, kote `iloc` la retire denye vale a, `loc` li menm enklisif pa rapo ak denye vale ou spesifye nan `range` ou an